### PR TITLE
Refine mask

### PR DESCRIPTION
**Pull request by CPFLAME**
这个PR要做的 
- [x] 优化mask的数据类型, int64 -> bool

和main分支进行了loss对齐, 验证正确性没有问题

加了输出信息看了一下T5中的数据格式, 也符合预期:
```python
1: lm_labels, oneflow.Size([16, 128]), oneflow.int64
1: loss_mask, oneflow.Size([16, 128]), oneflow.bool
1: encoder_input_ids, oneflow.Size([16, 512]), oneflow.int64
1: decoder_input_ids, oneflow.Size([16, 128]), oneflow.int64
1: encoder_attn_mask, oneflow.Size([16, 512, 512]), oneflow.bool
1: decoder_attn_mask, oneflow.Size([16, 128, 128]), oneflow.bool
1: encoder_decoder_attn_mask, oneflow.Size([16, 128, 512]), oneflow.bool
```

### bert loss对齐

以`configs/bert_large_pretrain.py`为例, 
在configs中加上以下配置, 
并把`libai/data/build.py`中`persistent_workers=True if num_workers > 0 else False,`全部注释掉, 
```python
train.output_dir = "output/bert_output_loss_compare"
train.load_weight = "../libai/output/bert_output/model_final/" # 采用同一个model进行初始化
train.rdma_enabled = False
model.cfg.hidden_dropout_prob= 0.0 # 关闭所有的dropout
model.cfg.attention_probs_dropout_prob= 0.0
model.cfg.bias_dropout_fusion= False
train.train_iter = 100
train.dist.pipeline_parallel_size=2
train.evaluation.enabled = False
```

最后loss对齐的曲线为:
image

### t5 loss对齐

以`configs/t5_large_pretrain.py`为例, 
在configs中加上一下配置,
并把`libai/data/build.py`中`persistent_workers=True if num_workers > 0 else False,`全部注释掉, 
```python
train.output_dir = "output/t5_output_loss_compare"
train.load_weight = "../libai/output/t5_output/model_final/" # 采用同一个model进行初始化
train.rdma_enabled = False
model.cfg.hidden_dropout_prob= 0.0 # 关闭所有的dropout
model.cfg.attention_probs_dropout_prob= 0.0
model.cfg.embedding_dropout_prob=0.0
model.cfg.bias_dropout_fusion= False
train.train_iter = 100
train.dist.pipeline_parallel_size=2
train.evaluation.enabled = False
```

最后loss曲线为
image

### CODE REVIEW

**Review by chengtbf — APPROVED** [review]

**Review by xiezipeng-ML — APPROVED** [review]

**Comment by CPFLAME** [comment]
@chengtbf   这个pr需要等libai发版以后再合吗?

**Comment by xiezipeng-ML** [comment]
T5里的这个也可以改
```
class ExtendedMask(flow.nn.Module):
    def forward(self, attention_mask):
        extended_attention_mask = attention_mask.unsqueeze(1)
        # Convert attention mask to binary.
        extended_attention_mask = extended_attention_mask > 0.5
        return extended_attention_mask
```

**Comment by chengtbf** [comment]
> @chengtbf 这个pr需要等libai发版以后再合吗?

可以现在合并。应该对内存有一些改善。

**Comment by CPFLAME** [comment]
LOSS 对齐注意的事项,  除了上述的修改以外

以`configs/bert_large_pretrain.py`为例, 
在configs中加上以下配置, 
并把`libai/data/build.py`中`persistent_workers=True if num_workers > 0 else False,`全部注释掉, 
```python
train.output_dir = "output/bert_output_loss_compare"
train.load_weight = "../libai/output/bert_output/model_final/" # 采用同一个model进行初始化
train.rdma_enabled = False
model.cfg.hidden_dropout_prob= 0.0 # 关闭所有的dropout
model.cfg.attention_probs_dropout_prob= 0.0
model.cfg.bias_dropout_fusion= False
train.train_iter = 100
train.dist.pipeline_parallel_size=2
train.evaluation.enabled = False
```

还需要在libai里面额外添加几行代码, 保存每个iter的loss.
在 https://github.com/Oneflow-Inc/libai/blob/7c22aa1bdb6e2a44c774149274a0290687912183/libai/engine/trainer.py#L213 这一行下面加上保存loss.txt的语句:

```python
total_losses_reduced = sum(metrics_dict.values())
if dist.is_main_process():
    txt = open("your_loss.txt", "a")
    txt.write(str(total_losses_reduced.item())+"\n")
```
因为是用`a`进行append追加, 所以如果要重新跑的话,  记得先把`your_loss.txt`删除

显示loss曲线的python脚本为:
```python
import os
import argparse
import matplotlib.pyplot as plt
import numpy as np


def _parse_args():
    parser = argparse.ArgumentParser("flags for draw image")
    parser.add_argument(
        "--txt_root",
        type=str,
        default="./results/check_info/",
        help="your txt root dir",
    )
    parser.add_argument(
        "--save_root", type=str, default="./loss_txt/", help="your draw image save dir",
    )
    return parser.parse_args()


# helpers
def load_data(file_path):
    def load_txt(file_path):
        data = []
        with open(file_path, "r") as f:
            for _line in f.readlines():
                data.append(float(_line.strip()))
        return data

    if type(file_path) == type([]):
        total = 0
        for fp in file_path:
            total += np.array(load_txt(fp))
        total = total / len(file_path)
        return total.tolist()
    else:
        return load_txt(file_path)


def draw_and_save(info_dic):
    # info_dic: {
    #   "title": "compare_loss"
    #   "save_path": "your_save_path"
    #   "txts": [a.txt, b.txt],
    #   "names": [a_name, b_name],
    #   "xlabel": "epochs",
    #   "ylabel": "acc",
    #   "xlim": [0, 1],                  # Optional
    #   "ylim": [0, 1],                  # Optional
    #   "do_abs_minus": False            # Optional
    # }
    title, save_path = info_dic["title"], info_dic["save_path"]
    txts, labels = info_dic["txts"], info_dic["names"]
    xlabel, ylabel = info_dic["xlabel"], info_dic["ylabel"]
    xlim, ylim = info_dic.get("xlim", 0), info_dic.get("ylim", 0)
    do_abs_minus = info_dic.get("do_abs_minus", False)
    assert len(txts) == len(labels)

    # setup
    plt.rcParams["figure.dpi"] = 100
    plt.clf()
    plt.xlabel(xlabel, fontproperties="Times New Roman")
    plt.ylabel(ylabel, fontproperties="Times New Roman")
    if xlim:
        plt.xlim(xlim[0], xlim[1])
    if ylim:
        plt.ylim(ylim[0], ylim[1])

    if do_abs_minus:
        assert len(txts) == 2
        data1, data2 = load_data(txts[0]), load_data(txts[1])
        assert len(data1) == len(data2)
        idxs = [i for i in range(len(data1))]
        abs_data = [abs(data1[i] - data2[i]) for i in idxs]
        plt.plot(idxs, abs_data)
    else:
        for txt, label in zip(txts, labels):
            data = load_data(txt)[:300]
            idxs = [i for i in range(len(data))]
            plt.plot(idxs, data, label=label)
    plt.title(title)
    plt.legend(loc="upper right", frameon=True, fontsize=8)
    plt.savefig(save_path)


def add_pth(a, b):
    return os.path.join(a, b)


if __name__ == "__main__":
    args = _parse_args()
    # txt_root = args.txt_root
    save_root = args.save_root
    # assert os.path.exists(txt_root), 'you should run "check/check.sh" before drawing graphs'

    # draw and save
    os.makedirs(save_root, exist_ok=True)
 
    draw_and_save(
        {
            "title": "loss_curve_bench_vs_main_dp",
            "save_path": add_pth(".", "bench_vs_main_dp.png"),
            "txts": [
                "main_dp4_bert_loss.txt",
                "../libai/bench_dp4_bert_loss.txt",
            ],
            "names": [
                "main_dp4",
                "bench_dp4",
            ],  # "lazy_reapeat4part_loss"
            "xlabel": "iter",
            "ylabel": "loss",
        }
    )
```
